### DISCUSSION TITLE
creating-a-quartz-net-hosted-service-with-asp-net-core/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#124] [Announcements]
# creating-a-quartz-net-hosted-service-with-asp-net-core/

In this post I describe how to run Quartz.NET jobs using an ASP.NET Core hosted service, and how to handle scoped services inside singleton jobs.

https://andrewlock.net/creating-a-quartz-net-hosted-service-with-asp-net-core/

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at May 21 2019, 03:07, in reply to [Thomas Levesque](https://disqus.com/by/thomaslevesque/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/124#discussioncomment-6056514)</em>

---

<p>Good point, thanks, that should be Singleton or Transient (i.e. not scoped). I find I rarely use transients hence the mistake!</p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at May 21 2019, 06:42, in reply to [Thomas Levesque](https://disqus.com/by/thomaslevesque/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/124#discussioncomment-6056519)</em>

---

<p>Hmmm, while technically true, you still need to use the construction at the end of the post (manually creating a scope), otherwise your "scoped" services will actually be singletons (as they're created from the "root" container). The scoped services also won't be disposed.</p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at May 23 2019, 07:46, in reply to [Thomas Levesque](https://disqus.com/by/thomaslevesque/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/124#discussioncomment-6056523)</em>

---

<p><a href="https://disqus.com/by/InversionOfControl/">Matthew Abbott</a> uses that approach as in this gist: <a href="https://gist.github.com/Antaris/9c3c097d31a90da279e3e6d78497a369" rel="nofollow noopener" title="https://gist.github.com/Antaris/9c3c097d31a90da279e3e6d78497a369">https://gist.github.com/Antaris/9c3c097d31a90da279e3e6d78497a369</a></p><p>Unfortunately the design of the IJobFactory API makes it a bit clunky, but it's definitely an option :)</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at May 29 2019, 09:10, in reply to [k3davis](https://disqus.com/by/k3davis/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/124#discussioncomment-6056526)</em>

---

<p>To be honest, I've only ever used Quartz, so I can't comment :)</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at July 20 2019, 01:15, in reply to [Tobi Schweiger](https://disqus.com/by/tobischweiger/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/124#discussioncomment-6056535)</em>

---

<p>Glad it's useful. RE adding and removing triggers - I'm not entirely sure how that would fit in, as it's not something I've needed to do to be honest. I've always had a fixed set of jobs to run - any variation in the job is handled inside the IJob, instead of adding/removing triggers.</p><p>However, I expect you should be able to inject the ISchedulerFactory into your module, call GetScheduler(), and then manipulate triggers from there. You have shared mutable state at that point though - I'm not sure about the thread safety of ISchedulerFactory (I suspect it's not) so be careful (add locking where necessary etc)!</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at July 10 2020, 09:01, in reply to [Marko Lahma](https://disqus.com/by/markolahma/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/124#discussioncomment-6056584)</em>

---

<p>That's awesome! I actually just noticed your GitHub pings, apologies. I've been taking a break from that side at the moment, and didn't spot it. I'll try and take a look through soon. Thanks for all your work on this 🙂</p>

[a comment by an automated account is not shown]
